### PR TITLE
[WIP] Add `qml.state()` support to `lightning.tensor` 

### PR DESCRIPTION
**Pull request by multiphaseCFD**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [ ] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane_lightning/lightning_tensor/lightning_tensor.py [hunk 1]
```diff
@@ -117,8 +117,6 @@
         "OrbitalRotation",
         "QFT",
         "ECR",
-        "BlockEncode",
-        "C(BlockEncode)",
```
Comment (vincentmr) [inline pennylane_lightning/lightning_tensor/lightning_tensor.py:121]: Why is this removed?

File: tests/new_api/test_device.py [hunk 2]
```diff
@@ -383,26 +391,38 @@ def test_preprocess_state_prep_middle_op_decomposition(self, op, decomp_depth):
...
+                    qml.SparseHamiltonian(
+                        qml.Hamiltonian([-1.0, 1.5], [qml.Z(1), qml.X(1)]).sparse_matrix(
+                            wire_order=[0, 1, 2]
+                        ),
+                        wires=[0, 1, 2],
+                    )
+                ),
+                qml.expval(qml.Projector([1], wires=2)),
+            ]
+            if device_name != "lightning.tensor"
```
Comment (vincentmr) [inline tests/new_api/test_device.py:415]: Could we instead use `pytest.skip` in the test and filter unsupported operations?

File: tests/new_api/test_device.py [hunk 3]
```diff
@@ -426,21 +446,37 @@ def test_execute_single_measurement(self, theta, phi, mp, dev):
...
-            qml.var(qml.Hermitian(qml.Hadamard.compute_matrix(), 0)),
-        ],
+        (
+            [
+                qml.probs(wires=[1, 2]),
+                qml.expval(qml.Z(2)),
+                qml.var(qml.X(2)),
+                qml.var(qml.Hermitian(qml.Hadamard.compute_matrix(), 0)),
+            ]
+            if device_name != "lightning.tensor"
```
Comment (vincentmr) [inline tests/new_api/test_device.py:456]: Idem.

File: tests/test_templates.py [hunk 4]
```diff
@@ -98,7 +98,7 @@ def test_angleembedding(self, n_qubits):
 
         def circuit(feature_vector):
             qml.AngleEmbedding(features=feature_vector, wires=range(n_qubits), rotation="Z")
-            return qml.state() if device_name != "lightning.tensor" else qml.expval(qml.PauliZ(0))
+            return qml.state()
```
Comment (vincentmr) [inline tests/test_templates.py:101]: Nice.

**Review by vincentmr — COMMENTED** [review]
Nice to see this PR @multiphaseCFD . I'm leaving a few suggestions. Please ping me again when you would like another round of reviews.

File: pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/tests/Tests_MPSTNCuda.cpp [hunk 5]
```diff
@@ -178,4 +177,17 @@ TEMPLATE_TEST_CASE("MPSTNCuda::getDataVector()", "[MPSTNCuda]", float, double) {
         CHECK(expected_state ==
               Pennylane::Util::approx(mps_state.getDataVector()));
     }
+
+    SECTION("Throw error for getDataVector() on device") {
+        std::size_t num_qubits = 100;
```
Comment (maliasadi) [inline pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/tests/Tests_MPSTNCuda.cpp:182]: Would you add tests for zero number of qubits?

File: pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/tests/Tests_MPSTNCuda.cpp [hunk 6]
```diff
@@ -178,4 +177,17 @@ TEMPLATE_TEST_CASE("MPSTNCuda::getDataVector()", "[MPSTNCuda]", float, double) {
...
+        std::size_t maxBondDim = 2;
+        DevTag<int> dev_tag{0, 0};
+
+        MPSTNCuda<TestType> mps_state{num_qubits, maxBondDim, dev_tag};
+
+        REQUIRE_THROWS_WITH(
+            mps_state.getDataVector(),
+            Catch::Matchers::Contains(
+                "State tensor size exceeds the available GPU memory!"));
```
Comment (maliasadi) [inline pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/tests/Tests_MPSTNCuda.cpp:191]: Thanks! :slightly_smiling_face:

File: tests/new_api/test_device.py [hunk 7]
```diff
@@ -324,13 +326,19 @@ def test_preprocess(self, adjoint):
...
-            (qml.BasisState([1, 1], wires=[0, 1]), False),
-            (qml.BasisState(qml.numpy.array([1, 1]), wires=[0, 1]), True),
-        ],
+        (
+            [
+                (qml.StatePrep([1 / np.sqrt(2), 1 / np.sqrt(2)], wires=0), False),
+                (qml.StatePrep(qml.numpy.array([1 / np.sqrt(2), 1 / np.sqrt(2)]), wires=0), True),
+                (qml.StatePrep(np.array([1, 0]), wires=0), False),
+                (qml.BasisState([1, 1], wires=[0, 1]), False),
+                (qml.BasisState(qml.numpy.array([1, 1]), wires=[0, 1]), True),
```
Comment (maliasadi) [inline tests/new_api/test_device.py:335]: We decompose them right now. right?

File: pennylane_lightning/lightning_tensor/_measurements.py [hunk 8]
```diff
@@ -95,7 +114,9 @@ def var(self, measurementprocess: MeasurementProcess):
             Variance of the observable
         """
         if isinstance(measurementprocess.obs, qml.SparseHamiltonian):
-            raise NotImplementedError("Sparse Hamiltonian Observables are not supported.")
+            raise NotImplementedError(
```
Comment (maliasadi) [inline pennylane_lightning/lightning_tensor/_measurements.py:117]: Where is test for this exception added?

File: pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/TNCudaBase.hpp [hunk 9]
```diff
@@ -293,6 +294,152 @@ class TNCudaBase : public TensornetBase<PrecisionT, Derived> {
...
+    /**
+     * @brief Get a slice of the state tensor
+     *
+     * @param tensor_data Pointer to the device memory for state tensor data.
+     * @param tensor_data_size Size of the state tensor data.
+     * @param wires Wires to get the state tensor for.
+     * @param numHyperSamples Number of hyper samples to use in the calculation
+     * and is default as 1.
+     */
+    void get_state_tensor(CFP_t *tensor_data,
```
[the remark this answers is not in the document]
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/TNCudaBase.hpp:350]: It will be called in the measurement class for prob support and let's keep as it.

File: pennylane_lightning/lightning_tensor/_measurements.py (hunk 8 above)
Comment (multiphaseCFD) [reply pennylane_lightning/lightning_tensor/_measurements.py:117]: yea, it is tested [here](https://github.com/PennyLaneAI/pennylane-lightning/blob/1c69f8fc3c9eafbb669b594f8afe961e0df1a43b/tests/lightning_tensor/test_gates_and_expval.py#L259-L263)

File: tests/lightning_tensor/test_gates_and_expval.py [hunk 10]
```diff
@@ -299,21 +298,20 @@ def test_measurement_shot_not_supported(self):
...
-        with pytest.raises(NotImplementedError, match="Shots are not supported for tensor network"):
+        with pytest.raises(
+            NotImplementedError, match="Shots are not supported for tensor network simulations."
+        ):
             m.measure_tensor_network(tape)
 
     def test_measurement_not_supported(self):
         """Test error for measure_tensor_network."""
-        obs = [qml.sample(wires=0)]
+        obs = [qml.sample(wires=[0])]
```
[the remark this answers is not in the document]
Comment (multiphaseCFD) [reply tests/lightning_tensor/test_gates_and_expval.py:308]: Nice catch. wires=0 works

File: tests/new_api/test_device.py (hunk 7 above)
Comment (multiphaseCFD) [reply tests/new_api/test_device.py:335]: Not really. Will work on the statevector to MPS decomposition after get all the measurement support

File: pennylane_lightning/lightning_tensor/lightning_tensor.py (hunk 1 above)
Comment (multiphaseCFD) [reply pennylane_lightning/lightning_tensor/lightning_tensor.py:121]: We only have limited support for BlockEncode and Controlled ones, as tested before. With more tests were in, we found BlockEncode are not fully supported. Will work on this to get all the gate support later when working on statevector to MPS transform.

File: doc/lightning_tensor/device.rst [hunk 11]
```diff
@@ -134,6 +133,8 @@ The "lightning.tensor" supports 1- and 2-wire gate operations and all other oper
     ~pennylane.DoubleExcitationMinus
     ~pennylane.DoubleExcitationPlus
     ~pennylane.GlobalPhase
+    ~pennylane.BlockEncode
+
```
[the remark this answers is not in the document]
Comment (multiphaseCFD) [reply doc/lightning_tensor/device.rst:137]: See above. Current implementation does not have the full BlockEncode support.

File: pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/tests/Tests_MPSTNCuda.cpp (hunk 5 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/tests/Tests_MPSTNCuda.cpp:182]: MPS won't work for 0 and 1 qubit cases with cutensornet. This check is added.

File: pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/TNCudaBase.hpp [hunk 12]
```diff
@@ -293,6 +294,152 @@ class TNCudaBase : public TensornetBase<PrecisionT, Derived> {
...
+     * @param numHyperSamples Number of hyper samples to use in the calculation
+     * and is default as 1.
+     *
+     * @return Full the state tensor on the host memory
+     */
+    auto get_state_tensor(const int32_t numHyperSamples = 1)
+        -> std::vector<ComplexT> {
+        std::vector<std::size_t> wires(BaseType::getNumQubits());
+        std::iota(wires.begin(), wires.end(), 0);
+        return get_state_tensor(wires, numHyperSamples);
```
[the remark this answers is not in the document]
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/TNCudaBase.hpp:314]: Let's keep it as it is now since we have to get `qml.probs(wires)` in the following PR.

File: pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/bindings/LTensorTNCudaBindings.hpp [hunk 13]
```diff
@@ -53,12 +53,30 @@ using TensorNetBackends =
...
+            "getState",
+            [](TensorNet &tensor_network, np_arr_c &state) {
+                py::buffer_info numpyArrayInfo = state.request();
+                auto *data_ptr =
+                    static_cast<std::complex<PrecisionT> *>(numpyArrayInfo.ptr);
+                if (state.size()) {
+                    auto &&state_tensor = tensor_network.getDataVector();
+                    std::copy(state_tensor.begin(), state_tensor.end(),
+                              data_ptr);
+                }
```
[the remark this answers is not in the document]
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/bindings/LTensorTNCudaBindings.hpp:77]: Nice one!

**Review by vincentmr — COMMENTED** [review]
Thanks @multiphaseCFD . LGTM, but I would tend to request adding a `getDataVector` function such as to reduce boilerplate in the unit tests (which could itself call `getData` though). I think we need both. What was the rationale for removing it?

File: pennylane_lightning/lightning_tensor/_tensornet.py [hunk 14]
```diff
@@ -67,6 +67,9 @@ def __init__(
         if device_name != "lightning.tensor":
             raise DeviceError(f'The device name "{device_name}" is not a valid option.')
 
+        if num_wires < 2:
+            raise ValueError("Number of wires must be greater than 1.")
```
[the remark this answers is not in the document]
Comment (multiphaseCFD) [reply pennylane_lightning/lightning_tensor/_tensornet.py:71]: correct, that's a limitation from backend.

**Comment by multiphaseCFD** [comment]
> Thanks @multiphaseCFD . LGTM, but I would tend to request adding a `getDataVector` function such as to reduce boilerplate in the unit tests (which could itself call `getData` though). I think we need both. What was the rationale for removing it?

Thanks @vincentmr ! The main reason is more product oriented. `getData()` will let our user to ensure that there is sufficient amount of host memory to store the state. It's the same logic of `cutensornet`, users have to be responsible for allocating sufficient amount of memory.  Changes were made in `lightning.tensor` since the lack of host memory issue could be more prominent for the 100+ qubit `MPS` or Exact `TN` simulations.

**Review by vincentmr — APPROVED** [review]
LGTM, nice work @multiphaseCFD .

File: pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/bindings/LTensorTNCudaBindings.hpp [hunk 15]
```diff
@@ -53,12 +53,27 @@ using TensorNetBackends =
...
+    using np_arr_c = py::array_t<std::complex<ParamT>,
+                                 py::array::c_style | py::array::forcecast>;
 
     pyclass
         .def(py::init<const std::size_t,
                       const std::size_t>()) // num_qubits, max_bond_dim
         .def(py::init<const std::size_t, const std::size_t,
                       DevTag<int>>()) // num_qubits, max_bond_dim, dev-tag
+        .def(
+            "getState",
```
Comment (maliasadi) [inline pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/bindings/LTensorTNCudaBindings.hpp:68]: Do we need any checks to catch data overflows in case the destination Numpy buffer is insufficient for the copy?

File: tests/test_comparison.py [hunk 16]
```diff
@@ -59,14 +59,14 @@ def one_qubit_block(wires=None):
...
-    device_name == "lightning.tensor",
-    reason="lightning.tensor device dose not support state return",
-)
 class TestComparison:
     """A test that compares the output states of the lightning device and ``default.qubit`` for a
     variety of different circuits. This uses ``default.qubit`` as a reference."""
 
+    @pytest.mark.skipif(
+        device_name == "lightning.tensor",
+        reason="lightning.tensor device dose not support 1 wires tensor network",
```
Comment (maliasadi) [inline tests/test_comparison.py:68]:
```suggestion
        reason="lightning.tensor device dose not support 1 wire tensor network",
```

File: tests/test_comparison.py [hunk 17]
```diff
@@ -102,10 +102,15 @@ def circuit(measurement):
         assert np.allclose(lightning_state, default_state)
         assert os.getenv("OMP_NUM_THREADS") == str(num_threads)
 
+    @pytest.mark.skipif(
+        device_name == "lightning.tensor",
+        reason="lightning.tensor device dose not direct access to state",
```
Comment (maliasadi) [inline tests/test_comparison.py:107]:
```suggestion
        reason="lightning.tensor device dose not support direct access to the state",
```

File: tests/test_comparison.py [hunk 18]
```diff
@@ -138,17 +143,22 @@ def circuit(measurement):
...
 
         default_state = default(qml.state)
 
+        # pylint: disable=protected-access
+        lightning_state = dev_l._statevector.state if dev_l._new_API else dev_l.state
         assert np.allclose(lightning_state, default_state)
 
+    @pytest.mark.skipif(
+        device_name == "lightning.tensor",
+        reason="lightning.tensor device dose not direct access to state",
```
Comment (maliasadi) [inline tests/test_comparison.py:155]:
```suggestion
        reason="lightning.tensor device dose not support the direct access to state",
```

File: tests/test_comparison.py [hunk 19]
```diff
@@ -189,17 +199,22 @@ def circuit(measurement):
...
```
Comment (maliasadi) [inline tests/test_comparison.py:211]:
```suggestion
        reason="lightning.tensor device dose not support the direct access to state",
```

File: tests/test_templates.py [hunk 20]
```diff
@@ -812,6 +802,10 @@ def circuit(basis_state):
 class TestQSVT:
     """Test the QSVT algorithm."""
 
+    @pytest.mark.skipif(
+        device_name == "lightning.tensor",
+        reason="lightning.tensor does not support BlockEncode",
```
Comment (maliasadi) [inline tests/test_templates.py:807]: We could provide support for BlockEncode by decomposing it to `qml.PrepSelPrep` in a followup PR.

**Review by maliasadi — APPROVED** [review]
Nice work @multiphaseCFD! LGTM if you accept the suggestions before merge :)

File: tests/test_templates.py (hunk 20 above)
Comment (multiphaseCFD) [reply tests/test_templates.py:807]: yea, will do it when working on all gate support.

File: pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/TNCudaBase.hpp [hunk 21]
```diff
@@ -294,6 +300,132 @@ class TNCudaBase : public TensornetBase<PrecisionT, Derived> {
...
+
+        PL_CUTENSORNET_IS_SUCCESS(cutensornetAccessorPrepare(
+            /* const cutensornetHandle_t */ getTNCudaHandle(),
+            /* cutensornetStateAccessor_t*/ accessor,
+            /* size_t */ scratchSize,
+            /* cutensornetWorkspaceDescriptor_t */ workDesc,
+            /* cudaStream_t unused as of v24.03 */ 0x0));
+
+        // Allocate workspace buffer
+        std::size_t worksize =
```
[the remark this answers is not in the document]
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/TNCudaBase.hpp:385]: Unfortunately, we can't set worksize as const here

File: pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/bindings/LTensorTNCudaBindings.hpp (hunk 15 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/bindings/LTensorTNCudaBindings.hpp:68]: yes, and it's checked inside the `getData()` API.

File: pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/TNCudaBase.hpp [hunk 22]
```diff
@@ -294,6 +300,132 @@ class TNCudaBase : public TensornetBase<PrecisionT, Derived> {
...
+            /* size_t */ sizeof(numHyperSamples)));
+
+        // prepare the computation
+        cutensornetWorkspaceDescriptor_t workDesc;
+        PL_CUTENSORNET_IS_SUCCESS(
+            cutensornetCreateWorkspaceDescriptor(getTNCudaHandle(), &workDesc));
+
+        // TODO we assign half (magic number is) of free memory size to the
+        // maximum memory usage.
+        const std::size_t scratchSize = cuUtil::getFreeMemorySize() / 2;
```
[the remark this answers is not in the document]
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/TNCudaBase.hpp:375]: Theoretically yes, but if the scratchSize is less 1 byte, the following call will fail.

File: tests/test_templates.py (hunk 20 above)
Comment (maliasadi) [reply tests/test_templates.py:807]: Awesome!
